keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,490 @@
+
+2. when each k8s Node is initialized with an environment file that is used by
+   Pods running on that node, k8s can provide a mechanism to instantiate env
+   vars from this file.
+
+3. The user is using a container offered by a vendor that requires configurating
+   environment variables (for eg, license key, one-time secret tokens). They can
+   use an `initContainer` to fetch this info into a file and the main container
+   can consume.
+
+The syntactic enhancement proposed in this KEP offers a simplified method for
+populating environment variables directly from files. This change would
+streamline the process for users.
+
+### Goals
+
+- Define container environment variables from a file.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+The user is using a container offered by a vendors that requires configuring
+env vars. [Some
+vendors](https://developer.hashicorp.com/vault/docs/platform/k8s/injector/examples#environment-variable-example)
+have to couple their app with Kubernetes to allow setting env vars in their
+container.
+
+Instead, k8s can provide a new mechanism that will be used to instantiate
+env vars from the specified file. The user can generate an env file using an
+`initContainer`. The container can then use the `fileRef` field to reference
+the file from which it should initialize the env vars. In the example below,
+the Pod's output includes `CONFIG_VAR=HELLO`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo CONFIG_VAR="HELLO" > /etc/config/config.env']
+    volumeMounts:
+    - name: data
+      mountPath: /etc/config
+  containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    envFrom:
+    - fileRef:
+        path: config.env
+        volumeName: data
+  restartPolicy: Never
+  volumes:
+  - name: data
+    emptyDir: {}
+```
+
+#### Story 2
+
+Each k8s Node is instantiated with an env file that contains information about
+the Node (for eg, instance type, failure zones). The user would like to
+provide some of the information in this file to the vendor's container via
+env vars. For example, consider that each node has the following env file:
+
+```
+CONFIG_VAR=HELLO
+CONFIG_VAR_A=WORLD
+...
+```
+
+The user can use the `fileKeyRef` field to select a subset of these env vars
+that it wants to provide to the app. In this case, the Pod's output includes
+`CONFIG_VAR=HELLO`
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_VAR
+      valueFrom:
+        fileKeyRef:
+          path: node.env
+          volumeName: data
+          key: CONFIG_VAR
+  restartPolicy: Never
+  volumes:
+  - name: data
+    hostPath: /etc/config
```

Comment:
what if host path will point to some file with the sensitive information?

I really feel that we can start this with empty-dir only. Expanding to host path seems to add a lot of security questions.

Reply:
The user will still be able to do weird things like creating an initContainer that copies a env file from hostPath to empty-dir and later the main container can consume this env file.

Reply:
Another thing to note here is that we won't be mounting the hostPath volume onto the main container. kubelet will just load envvars from the specified file in the hostPath during container creation. The main container will not have any access to the hostPath volume unless it is mounted for other stuff.

Reply:
> Another thing to note here is that we won't be mounting the hostPath volume onto the main container. kubelet will just load envvars from the specified file in the hostPath during container creation. The main container will not have any access to the hostPath volume unless it is mounted for other stuff.

Yes, this is the main concern. I also discussed it with @tallclair . There are tools in place which would control what is mounted to the Pod from the host. Adding this sneaky way to read files on the host will pass unnoticed with those tools. One option is to require to mount those and only allow to use files that are already mounted to the Pod. This will eliminate all security concerns